packages/jest-mock/src/index.ts

```diff
@@ -61,99 +59,106 @@ type ConstructorParameters<T> = T extends new (...args: infer P) => any
 export type MaybeMockedConstructor<T> = T extends new (
   ...args: Array<any>
 ) => infer R
-  ? MockInstance<R, ConstructorParameters<T>>
+  ? MockInstance<(...args: ConstructorParameters<T>) => R>
   : T;
 
-export interface MockWithArgs<T extends MethodLike>
-  extends MockInstance<ReturnType<T>, Parameters<T>> {
+export interface MockWithArgs<T extends FunctionLike>
+  extends MockInstance<FunctionType<T>> {
   new (...args: ConstructorParameters<T>): T;
   (...args: Parameters<T>): ReturnType<T>;
 }
 
-export type MockedFunction<T extends MethodLike> = MockWithArgs<T> & {
+export type MockedFunction<T extends FunctionLike> = MockWithArgs<T> & {
   [K in keyof T]: T[K];
 };
 
-export type MockedFunctionDeep<T extends MethodLike> = MockWithArgs<T> &
+export type MockedFunctionDeep<T extends FunctionLike> = MockWithArgs<T> &
   MockedObjectDeep<T>;
 
 export type MockedObject<T> = MaybeMockedConstructor<T> & {
-  [K in MethodLikeKeys<T>]: T[K] extends MethodLike
+  [K in MethodLikeKeys<T>]: T[K] extends FunctionLike
     ? MockedFunction<T[K]>
     : T[K];
 } & {[K in PropertyLikeKeys<T>]: T[K]};
 
 export type MockedObjectDeep<T> = MaybeMockedConstructor<T> & {
-  [K in MethodLikeKeys<T>]: T[K] extends MethodLike
+  [K in MethodLikeKeys<T>]: T[K] extends FunctionLike
     ? MockedFunctionDeep<T[K]>
     : T[K];
 } & {[K in PropertyLikeKeys<T>]: MaybeMockedDeep<T[K]>};
 
-export type MaybeMocked<T> = T extends MethodLike
+export type MaybeMocked<T> = T extends FunctionLike
   ? MockedFunction<T>
   : T extends object
   ? MockedObject<T>
   : T;
 
-export type MaybeMockedDeep<T> = T extends MethodLike
+export type MaybeMockedDeep<T> = T extends FunctionLike
   ? MockedFunctionDeep<T>
   : T extends object
   ? MockedObjectDeep<T>
   : T;
 
 export type Mocked<T> = {
-  [P in keyof T]: T[P] extends MethodLike
-    ? MockInstance<ReturnType<T[P]>, Parameters<T[P]>>
-    : T[P] extends ConstructorLike
+  [P in keyof T]: T[P] extends FunctionLike
+    ? MockInstance<FunctionType<T[P]>>
+    : T[P] extends ClassLike
     ? MockedClass<T[P]>
     : T[P];
 } & T;
 
-export type MockedClass<T extends ConstructorLike> = MockInstance<
-  InstanceType<T>,
-  T extends new (...args: infer P) => any ? P : never
+export type MockedClass<T extends ClassLike> = MockInstance<
+  (args: T extends new (...args: infer P) => any ? P : never) => InstanceType<T>
 > & {
   prototype: T extends {prototype: any} ? Mocked<T['prototype']> : never;
 } & T;
 
-export interface Mock<T, Y extends Array<unknown> = Array<unknown>>
+type UnknownFunction = (...args: Array<unknown>) => unknown;
+
+export interface Mock<T extends FunctionLike = UnknownFunction>
```

Comment:
There is a need to keep `any`s in `type FunctionLike = (...args: any) => any`. It does not work with `unknown`s either internally, or something breaks in tests. Despite `any`s, this is not `any` (; It is `AnyFunction` and that’s a constrain. Passing anything what is not a function will throw TS error. To avoid `any`s leak to user side I used defaults of `(...args: Array<unknown>) => unknown`. This means that using `fn()` without any implementation will default to `unknown`, if implementation is provided types are inferred.

Reply:
Trying to explain this in more detail, because I repeated this pattern several times. Internally all what’s necessary is to be sure that we have a function, hence simply any function works. It also works as a constrain. But the default on user side always is `(...args: Array<unknown>) => unknown`.